packages/use-contractkit/src/utils/useProviders.ts

```diff
@@ -0,0 +1,97 @@
+import { useMemo } from 'react';
+import { isMobile } from 'react-device-detect';
+
+import {
+  localStorageKeys,
+  Priorities,
+  PROVIDERS,
+  WalletTypes,
+} from '../constants';
+import { Maybe, Provider, WalletConnectProvider, WalletEntry } from '../types';
+import localStorage from './localStorage';
+import { defaultProviderSort } from './sort';
+
+export function walletToProvider(wallet: WalletEntry): WalletConnectProvider {
+  return {
+    name: wallet.name,
+    type: WalletTypes.WalletConnect,
+    description: wallet.description || 'Missing description in registry',
+    icon: wallet.logos.md,
+    canConnect: () => true,
+    showInList: () =>
+      isMobile ? Object.values(wallet.mobile).some(Boolean) : true,
+    listPriority: () => 0,
+    installURL: wallet.homepage,
+  };
+}
+
+export function getRecent(): Maybe<Provider> {
+  const type = localStorage.getItem(localStorageKeys.lastUsedWalletType);
+  const id = localStorage.getItem(localStorageKeys.lastUsedWalletId);
+  let provider;
+
+  if (id && WalletTypes.WalletConnect === type) {
+    provider = Object.values(PROVIDERS).find(
+      (p) => (p as WalletConnectProvider).walletConnectId === id
+    );
+  } else {
+    provider = Object.values(PROVIDERS).find((p) => p.type === type);
+  }
+
+  if (provider) return provider;
+  return null;
+}
+
+export default function useProviders(
+  wallets: WalletEntry[] = [],
+  sort = defaultProviderSort,
+  search?: string
+) {
+  const record: Record<string, Provider> = useMemo(
+    () => ({
+      ...PROVIDERS,
+      ...wallets.reduce((acc, wallet) => {
+        acc[wallet.id] = walletToProvider(wallet);
+        return acc;
+      }, {} as Record<string, Provider>),
+    }),
+    [wallets]
+  );
+
+  const providers = useMemo<[providerKey: string, provider: Provider][]>(() => {
+    let _record = Object.entries(record);
+    if (search) {
```

Comment:
This feels like it could be added to the filter below and then you wouldn't have to filter twice and also you wouldn't have to reassign `_record`

Reply:
Sure. Not that performance is the biggest issue here, was aiming for clarity rather than optimization. But I do agree, it can be there

Reply:
If typing in a search box that filters thru a list of rendered components ive found performance IS pretty important to avoid typing lag.